core/deadline_test.go

```diff
@@ -29,48 +31,87 @@ func TestDeadliner(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
-	const lateFactor = 1
-	startTime := time.Now()
+	expiredDuties, nonExpiredDuties, voluntaryExits, dutyExpired := setupData(t)
+	clock := clockwork.NewFakeClock()
 
-	deadlineFunc := func(duty core.Duty) time.Time {
-		duration := time.Second
-		if duty.Type == core.DutyExit {
-			// Do not timeout exit duties.
-			return time.Date(9999, 1, 1, 0, 0, 0, 0, time.UTC)
+	deadlineFuncProvider := func() func(duty core.Duty) time.Time {
+		startTime := clock.Now()
+		return func(duty core.Duty) time.Time {
+			if duty.Type == core.DutyExit {
+				return startTime.Add(time.Hour)
+			}
+
+			if dutyExpired(duty) {
+				return startTime.Add(-1 * time.Hour)
+			}
+
+			return startTime.Add(time.Duration(duty.Slot) * time.Second)
 		}
+	}
+
+	deadliner := core.NewForT(ctx, deadlineFuncProvider(), clock)
+
+	// Add our duties to the deadliner.
```

Comment:
will be adding a `sync.WaitGroup` to wait for these goroutines to finish adding their duties.

Reply:
Done!